[LIVY-396] Livy does not map YARN app states correctly

meisam opened the pull request:
Make sure that the invalid combinations of `state` and `finalStatus` do
not occur, and if they do, put the application in the `FAILED` state.

Task-Url: https://issues.apache.org/jira/browse/LIVY-396

jerryshao commented:
Looks like the UT is failed, @meisam can you please fix the UT.

meisam commented:
@jerryshao The test fails due to timing of the events. I cannot reproduce them on my local machine, but now that we have the travis logs, I'll work on fixing them. Do you want me to open a new JIRA ticket for it?

jerryshao commented:
If this fix to UT is just several lines, I think you can also fix which this PR.

ajbozarth commented:
This LGTM if you can address the travis failures

jerryshao commented:
@meisam , can you please fix the UT, thanks.

meisam commented:
The fix is coming with LIVY-336: #44

jerryshao commented:
I don't think we can merge this PR with UT failure, please fix this here.

meisam commented:
@jerryshao I updated the PR with fix to UTs.
Sorry for not being clear. What I meant was, I may close this PR without merging it because LIVY-336:#44 updates many unit tests, including the tests that fail here.
But on a second though, I see that getting LIVY-336 merged takes a while, so I went ahead and fixed tests here.

ajbozarth commented:
@meisam just an fyi when you amend then force push your updates it makes it hard to track changes made from feedback. In the future you can just make another commit to your branch and push it, then we can see what changes you made.

ajbozarth commented on server/src/test/scala/org/apache/livy/utils/SparkYarnAppSpec.scala:
```diff
@@ -58,21 +59,50 @@ class SparkYarnAppSpec extends FunSpec with LivyBaseUnitTestSuite {
         val mockYarnClient = mock[YarnClient]
         val mockAppListener = mock[SparkAppListener]
 
+
```
unnecessary extra blank line

meisam replied:
👍 I removed the empty line.

ajbozarth commented on server/src/test/scala/org/apache/livy/utils/SparkYarnAppSpec.scala:
```diff
@@ -58,21 +59,50 @@ class SparkYarnAppSpec extends FunSpec with LivyBaseUnitTestSuite {
-        // Simulate YARN app state progression.
-        when(mockAppReport.getYarnApplicationState).thenAnswer(new Answer[YarnApplicationState]() {
-          private var stateSeq = List(ACCEPTED, RUNNING, FINISHED)
+//        when(mockAppReport.getFinalApplicationStatus)
```
Why is this commented out line here?

meisam replied:
I replaced it a with `when(mockAppReport.getFinalApplicationStatus).thenAnswer( ...` a few lines down.

ajbozarth replied:
Then should it be removed instead of commented out?

meisam replied:
I removed it.

meisam commented:
> fyi when you amend then force push your updates it makes it hard to track changes made from feedback. In the future you can just make another commit to your branch and push it, then we can see what changes you made.
Thanks for the feedback. On feature/fix branches that are not merged, I rebase my own commits to keep the git commit history clean, but this is a personal preference.

ajbozarth commented:
> Thanks for the feedback. On feature/fix branches that are not merged, I rebase my own commits to keep the git commit history clean, but this is a personal preference.
I can completely understand that for projects were you commit to master directly, but in the case of Spark the merge script will squash all your commits and use the PR title and description for the commit message. So it's better in Spark work to not do rebases unless you need to merge master in multiple times (in which case that can get messy). This of course is all from a PR review perspective

jerryshao commented:
Looks like travis environment is changed, which makes python-api build fail, let me fix this issue first.

jerryshao commented:
@meisam would you please trigger the test again. I just fixed the travis issue.

meisam commented:
Closing and reopening the PR to trigger travis.

jerryshao commented:
LGTM, merging to master.
